heartbeat/monitors/wrappers/summarizer/summarizer.go

```diff
@@ -0,0 +1,173 @@
+	Status       monitorstate.StateStatus `json:"status"`
+	RetryGroup   string                   `json:"retry_group"`
+}
+
+func NewSummarizer(rootJob jobs.Job, sf stdfields.StdMonitorFields, mst *monitorstate.Tracker) *Summarizer {
+	uu, err := uuid.NewV1()
+	if err != nil {
+		logp.L().Errorf("could not create v1 UUID for retry group: %s", err)
+	}
+	return &Summarizer{
+		rootJob:        rootJob,
+		contsRemaining: 1,
+		mtx:            &sync.Mutex{},
+		jobSummary:     NewJobSummary(1, sf.MaxAttempts, uu.String()),
+		checkGroup:     uu.String(),
+		stateTracker:   mst,
+		sf:             sf,
+		// private property, but can be overriden in tests to speed them up
+		retryDelay: time.Second,
+	}
+}
+
+var id int
+
+func NewJobSummary(attempt uint16, maxAttempts uint16, retryGroup string) *JobSummary {
+	if maxAttempts < 1 {
+		maxAttempts = 1
+	}
+
+	id++
+	return &JobSummary{
+		Id:          id,
+		MaxAttempts: maxAttempts,
+		Attempt:     attempt,
+		RetryGroup:  retryGroup,
+	}
+}
+
+// Wrap wraps the given job in such a way that the last event summarizes all previous events
+// and additionally adds some common fields like monitor.check_group to all events.
+// This adds the state and summary top level fields.
+func (s *Summarizer) Wrap(j jobs.Job) jobs.Job {
+	return func(event *beat.Event) ([]jobs.Job, error) {
+		conts, jobErr := j(event)
+
+		_, _ = event.PutValue("monitor.check_group", s.checkGroup)
+
+		s.mtx.Lock()
+		defer s.mtx.Unlock()
+
+		js := s.jobSummary
+
+		s.contsRemaining-- // we just ran one cont, discount it
+		// these many still need to be processed
+		s.contsRemaining += uint16(len(conts))
+
+		monitorStatus, err := event.GetValue("monitor.status")
+		if err == nil && !eventext.IsEventCancelled(event) { // if this event contains a status...
+			mss := monitorstate.StateStatus(monitorStatus.(string))
+
+			if mss == monitorstate.StatusUp {
+				js.Up++
+			} else {
+				js.Down++
+			}
+		}
+
+		if s.contsRemaining == 0 {
+			if js.Down > 0 {
+				js.Status = monitorstate.StatusDown
+			} else {
+				js.Status = monitorstate.StatusUp
+			}
+
+			// Get the last status of this monitor, we use this later to
+			// determine if a retry is needed
+			lastStatus := s.stateTracker.GetCurrentStatus(s.sf)
+
+			// FinalAttempt is true if no retries will occur
+			js.FinalAttempt = js.Status != monitorstate.StatusDown || js.Attempt >= js.MaxAttempts
+
+			ms := s.stateTracker.RecordStatus(s.sf, js.Status, js.FinalAttempt)
+
+			eventext.MergeEventFields(event, mapstr.M{
+				"summary": js,
+				"state":   ms,
+			})
+
+			logp.L().Debugf("attempt info: %v == %v && %d < %d", js.Status, lastStatus, js.Attempt, js.MaxAttempts)
+			if !js.FinalAttempt {
+				// Reset the job summary for the next attempt
+				// We preserve `s` across attempts
+				s.jobSummary = NewJobSummary(js.Attempt+1, js.MaxAttempts, js.RetryGroup)
+				s.contsRemaining = 1
+				s.checkGroup = fmt.Sprintf("%s-%d", s.checkGroup, s.jobSummary.Attempt)
```

Comment:
I've noticed this will generate the following sequence:
`1 st: uuid() -> 2nd uuid()+"1" -> 2nd uuid()+"-1" +"-2 -> nth uuid()+..."-<n-1>" +"-<n>"`

Is that the expected format?

Reply:
Yes, my thinking is that the UUID is already unique enough, and that it makes debugging easier to have the sequences. Also, you could potentially use a prefix query as an optimization in the future